### ISSUE TITLE
`enum_variant_names` warns on enum variants with the same name as the type

### ISSUE THREAD

**Issue by lynzrand** [rust-lang/rust-clippy#16895]
### Summary

`clippy::enum_variant_names` warns when the variant have the same name as the type. These cases are usually intentional and should be ignored instead.

Related: https://github.com/rust-lang/rust-clippy/issues/8324

Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=91c013ebceacb1106660e5a8004f4237

### Lint Name

enum_variant_names

### Reproducer

I tried this code:

```rust
#[allow(unused)]
enum Type {
    Type,
    Builtin(PrimitiveType),
    Func,
}

#[allow(unused)]
enum PrimitiveType{
    Int,
    Float
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: variant name ends with the enum's name
 --> src/lib.rs:3:5
  |
3 |     Type,
  |     ^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#enum_variant_names
  = note: `#[warn(clippy::enum_variant_names)]` on by default
```

I expected to see this happen:

- (no error occurs)

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

**Comment by QEEK-AI** [comment]
**QEEK.AI Agent:** 

"Thanks for the report! This is indeed a common pattern that the lint should probably handle better".

**The pattern:**
```rust
enum Type {
    Type,        // The "base" or "default" variant
    Builtin(...),
    Func,
}
```

This is a well-established idiom in Rust — using the same name for a variant when it represents the "plain" or "default" case of that type. It's particularly common in:
- AST representations (`enum Expr { Expr, ... }`)
- Type system implementations (`enum Type { Type, ... }`)
- Parser/token enums

**Related context:**
As you noted, this connects to #8324 which discusses similar false positives. The lint currently triggers on any variant name that *ends with* the enum name, but an exact match is semantically different and typically intentional.

**Potential fixes:**

1. **Add an exception for exact matches** — Skip the lint when `variant_name == enum_name`

2. **Make it configurable** — Add a configuration option to allow exact name matches

3. **Downgrade the severity** — Change from `Warn` to `Allow` for exact matches (though this might be too subtle)

I lean toward **option 1** since exact matches are overwhelmingly intentional. The lint's purpose is to catch redundant naming like `enum Color { RedColor, BlueColor }`, not to discourage the `Type::Type` pattern.

**Would you be interested in submitting a PR for this? The relevant code is in `clippy_lints/src/enum_variants.rs` around where it checks `ends_with()`. We could add a simple check to skip when the names are identical.**

**Comment by lynzrand** [comment]
ok why is there an AI bot spamming?

Anyway, to anyone that haven't read enough context, I have already submitted a fix PR at https://github.com/rust-lang/rust-clippy/pull/16896. If you are a maintainer of this repo, please read that PR too.

**Comment by lmmx** [comment]
This bot has spammed 6+ issues in this repo, extremely not cool
